probe_src/performance_test.py

```diff
@@ -1,200 +1,174 @@
 import subprocess
 import datetime
 import csv
-import psutil
 import time
+import os
+import shutil
+import resource
 from dataclasses import dataclass
+import psutil
+import errno
 
 @dataclass
 class Result:
     returncode: int
-    cpu_times: tuple
-    memory_info: tuple
-    io_counters: tuple
     stdout: str
     stderr: str
-    start_time: datetime.datetime
-    end_time: datetime.datetime
-
-def benchmark_command(command: str, warmup_iterations: int, benchmark_iterations: int) -> list[Result]:
+    duration: float
+    rusage: resource.struct_rusage
+
+LOG_FILE = "probe_log"
+RECORD_DIR = "probe_record"
+
+class ResourcePopen(subprocess.Popen):
+    def _try_wait(self, wait_flags):
+        try:
+            (pid, sts, res) = os.wait4(self.pid, wait_flags)
+        except OSError as e:
+            if e.errno != errno.ECHILD:
+                raise
+            pid = self.pid
+            sts = 0
+        else:
+            self.rusage = res
+        return (pid, sts)
+
+def resource_call(*popenargs, timeout=None, **kwargs):
+    with ResourcePopen(*popenargs, **kwargs) as p:
+        try:
+            stdout, stderr = p.communicate(timeout=timeout)
+            return p.returncode, p.rusage, stdout, stderr
+        except:
+            p.kill()
+            stdout, stderr = p.communicate()
+            raise
+
```

Comment:
```suggestion
def resource_call(*popenargs, timeout=None, **kwargs) -> Result:
    with ResourcePopen(*popenargs, stdout=subprocess.PIPE, stderr=subprocess.PIPE, **kwargs) as p:
        start = datetime.datetime.now()
        try:
            stdout, stderr = p.communicate(timeout=timeout)
        except:
            p.kill()
            stdout, stderr = p.communicate()
            raise
        stop = datetime.datetime.now()
        return Result(p.returncode, stdout.decode(), stderr.decode(), (stop - start).total_seconds(), p.rusage)
```

This could eliminate some duplicate code, since we always need the duration, the rusage, and to construct a Result.